snow/engine/common/timer.go

```diff
@@ -3,12 +3,113 @@
 
 package common
 
-import "time"
+import (
+	"sync"
+	"time"
+)
 
-// Timer describes the standard interface for specifying a timeout
-type Timer interface {
+// PreemptionSignal signals when to preempt the pendingTimeout of the timeout handler.
+type PreemptionSignal struct {
+	activateOnce sync.Once
+	initOnce     sync.Once
+	signal       chan struct{}
+}
+
+func (ps *PreemptionSignal) init() {
+	ps.signal = make(chan struct{})
+}
+
+// Listen returns a read-only channel that is closed when Preempt() is invoked.
+func (ps *PreemptionSignal) Listen() <-chan struct{} {
+	ps.initOnce.Do(ps.init)
+	return ps.signal
+}
+
+// Preempt causes any past and future calls of Listen to return a closed channel.
+func (ps *PreemptionSignal) Preempt() {
+	ps.initOnce.Do(ps.init)
+	ps.activateOnce.Do(func() {
+		close(ps.signal)
+	})
+}
+
+// timeoutScheduler schedules timeouts to be dispatched in the future.
+// Only a single timeout can be pending to be scheduled at any given time.
+// Once a preemption signal is closed, all timeouts are immediately dispatched.
+type timeoutScheduler struct {
+	newTimer         func(duration time.Duration) *time.Timer
+	onTimeout        func()
+	preemptionSignal <-chan struct{}
+	pendingTimeout   chan struct{}
+}
+
+// NewTimeoutScheduler constructs a new timeout scheduler with the given function to be invoked upon a timeout,
+// unless the preemptionSignal is closed and in which case it invokes the function immediately.
+func NewTimeoutScheduler(onTimeout func(), preemptionSignal <-chan struct{}, newTimer func(duration time.Duration) *time.Timer) *timeoutScheduler {
+	pendingTimout := make(chan struct{}, 1)
+	pendingTimout <- struct{}{}
+	return &timeoutScheduler{
+		preemptionSignal: preemptionSignal,
+		newTimer:         newTimer,
+		onTimeout:        onTimeout,
+		pendingTimeout:   pendingTimout,
+	}
+}
+
+// RegisterTimeout fires the function the timeout scheduler is initialized with no later than the given timeout.
+func (th *timeoutScheduler) RegisterTimeout(d time.Duration) {
+	acquiredToken := th.acquirePendingTimeoutToken()
+	preempted := th.preempted()
+
+	if !preempted && !acquiredToken {
+		return
+	}
+
+	go th.scheduleTimeout(d, acquiredToken)
+}
+
+func (th *timeoutScheduler) scheduleTimeout(d time.Duration, acquiredToken bool) {
+	timer := th.newTimer(d)
+	defer timer.Stop()
+
+	defer th.onTimeout()
```

Comment:
nit: I feel like this makes a bit more sense to just be called inline at the end of the function.

Reply:
inlined

Reply:
I think this inlining may need to be pushed from your local branch, it still seems to use the `defer`.

Reply:
oh I thought you meant to inline relinquishing the pending token.